### PR TITLE
Pcid fixes

### PR DESCRIPTION
**Pull request by dsemenets**
Corrected error checking in function vchan_find_server

Signed-off-by: Dmitry Semenets <dmitry.semenets@gmail.com>

### CODE REVIEW

File: tools/libs/light/libxl_pcid.c [hunk 1]
```diff
@@ -854,6 +854,49 @@ static void server_fini_one(libxl__gc *gc, struct vchan_client *client)
...
+{
+    GC_INIT(ctx);
+    char *xs_path, *str;
+    char **dir = NULL;
+    unsigned int nb, i;
+    libxl_domid domid;
+    int res;
+    bool ret = false;
+
+
```
Comment (lorc) [inline tools/libs/light/libxl_pcid.c]: One empty line is enough

File: tools/libs/light/libxl_pcid.c [hunk 2]
```diff
@@ -854,6 +854,49 @@ static void server_fini_one(libxl__gc *gc, struct vchan_client *client)
...
+    pthread_mutex_init(&vchan_client_mutex, NULL);
```
Comment (lorc) [inline tools/libs/light/libxl_pcid.c]: It is strange to initialize mutex in such function. 

I believe, you might want to take it there, not init.

File: tools/libs/light/libxl_pcid.c [hunk 3]
```diff
@@ -1052,7 +1095,6 @@ int libxl_pcid_process(libxl_ctx *ctx)
          */
         if (!strcmp(watch_ret[XS_WATCH_PATH], xs_path))
             continue;
-
```
Comment (lorc) [inline tools/libs/light/libxl_pcid.c]: Please avoid unnecessary changes

**Review by lorc — COMMENTED** [review]
Thank you for the descriptive commit description. 

> When client closes connection pcid removes appropriate directory in xenstore.
After this xs_watch detects that directory changed and this caused false
detection of new client.

May we just remove watch before removing directory from xenstore? In this case, watch will not fire.


Also, could you please add at least one sentence to the other commits? 
You need to justify why are you removing `static` in the first commit. As for memoryleak fixes, you at least can write which functions are fixed...

File: tools/libs/light/libxl_pcid.c (hunk 1 above)
Comment (dsemenets) [reply tools/libs/light/libxl_pcid.c]: Ok

File: tools/libs/light/libxl_pcid.c (hunk 2 above)
Comment (dsemenets) [reply tools/libs/light/libxl_pcid.c]: Yes. It is mistake

File: tools/libs/light/libxl_pcid.c (hunk 3 above)
Comment (dsemenets) [reply tools/libs/light/libxl_pcid.c]: Done

**Comment by dsemenets** [comment]
> May we just remove watch before removing directory from xenstore? In this case, watch will not fire.

Remove directory and read_watch caused in different threads.

File: tools/libs/light/libxl_pcid.c [hunk 4]
```diff
@@ -854,6 +854,47 @@ static void server_fini_one(libxl__gc *gc, struct vchan_client *client)
     free(client);
 }
 
+static bool is_vchan_exist(libxl_ctx *ctx, char *watch_dir)
```
Comment (lorc) [inline tools/libs/light/libxl_pcid.c:858]: This is too complex. Lots of code is copied from the main function.

I played a bit with this. It is much easier to check if `watch_dir` exists. 

Because this is what you are trying to solve: watch fires for a deleted directory, so you can check if directory exists (with call to `xs_directory(.., watch_dir,...)`) and that's all.

File: tools/libs/light/libxl_pcid.c [hunk 5]
```diff
@@ -854,6 +854,30 @@ static void server_fini_one(libxl__gc *gc, struct vchan_client *client)
     free(client);
 }
 
+static bool is_vchan_exist(libxl_ctx *ctx, char *xs_path, char *watch_dir)
+{
+    char **dir = NULL;
+    unsigned int nb, i;
+    bool ret = false;
+    GC_INIT(ctx);
```
Comment (lorc) [inline tools/libs/light/libxl_pcid.c]: I don't think you need to init ctx again there. I afraid this can cause memory leak.

You can use the context that you got from the calling function.

File: tools/libs/light/libxl_pcid.c [hunk 6]
```diff
@@ -854,6 +854,30 @@ static void server_fini_one(libxl__gc *gc, struct vchan_client *client)
...
+    }
+    for (i=0; i<nb; i++) {
+        if (strstr(watch_dir, dir[i])) {
+            ret = true;
+            break;
+        }
+    }
+    free(dir);
+out:
+    GC_FREE;
```
Comment (lorc) [inline tools/libs/light/libxl_pcid.c]: If `GC_INIT` is not called, you don't need to call `GC_FREE`.

File: tools/libs/light/libxl_pcid.c [hunk 7]
```diff
@@ -854,6 +854,30 @@ static void server_fini_one(libxl__gc *gc, struct vchan_client *client)
...
 }
 
+static bool is_vchan_exist(libxl_ctx *ctx, char *xs_path, char *watch_dir)
+{
+    char **dir = NULL;
+    unsigned int nb, i;
+    bool ret = false;
+    GC_INIT(ctx);
+
+    dir = xs_directory(ctx->xsh, XBT_NULL, xs_path, &nb);
```
Comment (lorc) [inline tools/libs/light/libxl_pcid.c]: You can call `xs_directory(ctx->xsh, XBT_NULL, watch_dir, &nb)`. It will return `NULL` if `watch_dir` does not exist.

So, you will not need `for` loop below.

File: tools/libs/light/libxl_pcid.c (hunk 5 above)
Comment (dsemenets) [reply tools/libs/light/libxl_pcid.c:863]: Here is garbage collector init. Not context.

File: tools/libs/light/libxl_pcid.c (hunk 6 above)
Comment (dsemenets) [reply tools/libs/light/libxl_pcid.c:878]: Look above

File: tools/libs/light/libxl_pcid.c (hunk 7 above)
Comment (dsemenets) [reply tools/libs/light/libxl_pcid.c]: Done

**Comment by lorc** [comment]
`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`
